course_contents/2_intro_to_python/notes/02. Loops.md

```diff
@@ -0,0 +1,159 @@
+- if not specified, Python will implicitly set `start` to 0 and `step` to 1
+
+An example of the `range()` method used in a `for` loop:
+
+```py
+for i in range(5): #the range() method can be used to specify number of iterations of a loop
+  print(i) #the numbers printed will be: 0,1,2,3,4
+```
+
+### The `break` statement
+- the `break` statement is used to stop the loop from running when an external condition is reached
+- after the loop has been exited - the next line of code after it will be executed
+
+An example of using the `break` statement in the `while` loop:
+
+```py
+counter = 0
+while counter < 10:
+  print(counter)
+  if counter == 5:
+    break #the loop stops running when the counter is equal to 5
+  counter += 1
+print(f"That's all the numbers from 0 to {counter}!") #after the break, this line is executed
+```
+
+### The `continue` statement
+- the `continue` statement is used to skip a step in a loop and run the next iteration
+- no code after the `continue` statement will be run
+
+An example of the `continue` statement being used in a `for` loop:
+
+```py
+for i in range(5):
+    if i % 2 == 0: 
+        continue #if i is even, it won't be printed out
+    print(i)
+```
+
+### The `pass` statement
+- the `pass` statement is used to ignore a place in a code and continue running as if it doesn't exist
+- it is usually used as a placeholder at a place where some code is required from the syntax side, but the code doesn't exist yet
+
+An example of the `pass` statement being used in a `for` loop:
+
+```py
+for i in range(5):
+    if i % 2 == 0: 
+        pass #even if the block is entered nothing will happen because of the pass statement
+    print(i)
+```
+
+### `else` keyword and `for` loop
+- the `else` keyword can be used with `for` loops to check if the loop has finished running without breaking
+- if the loop hasn't been stopped by `break` the else block will be entered
+- the `else` keyword should be indented by the same number of spaces as `for` loop
+
+An example of using the `else` keyword with a `for` loop:
+
+```py
+for i in range(10):
+    print(i)
+    if i == 10:
+        break
+else:
+    print("The loop didn't break!")
+```
+
+An example of how not to use the `else` keyword with a `for` loop if you want to achieve the same goal as above:
+
+```py
+for i in range(10):
+    print(i)
+    if i == 10:
+        break
+    else:
+        print("The loop didn't break!")
+```
+
+### Nested loops:
+- loops can be used one inside another, be it `for` or `while` loops, you can mix them in any way
+- nested loops should be avoided when you aim for lower complexity of a program
+
+An example of nested loops in which the matrix of user entered dimensions is created:
+
+```py
+columns = int(input("Enter the number of columns: "))
+rows = int(input("Enter the number of rows: "))
+matrix = []
+for i in range(columns):
+  row = []
+  for j in range(rows):
+    row.append(0)
+  matrix.append(row)
+print(matrix)
```

Comment:
Could you come with a different example for this? As a reader, I wouldn't like to read a 1985 C code university style with variable like `i` and `j`.

Reply:
Variables i and j are usually used in maths when talking about matrices, and they are also used in a lot of programming books that are tackling the loops, so that's why I went for it, more because of tradition than making a point with their names.